### PR TITLE
Make finding URLs customizable

### PR DESCRIPTION
**Pull request by encukou**
The functionality is not written yet.

Fixes: https://github.com/encukou/freezeyt/issues/186

* [x] Documentation
* [x] Tests
* [x] Functionality

### CODE REVIEW

**Comment by nappex** [comment]
What means 
> A generic version of collections.abc.Iterator

see
https://docs.python.org/3/library/typing.html#typing.Iterator

**Comment by nappex** [comment]
why do we use var `file_path`, when we don't use it in code ?

see: https://github.com/encukou/freezeyt/pull/187/files#diff-210261fd8b8aa39c7026f85b96a4c5606de023aa2afa8a256995d807a2b0812aR273

It is good to know what we get when we read the code, but unused var should be name as `_`. So I little confused what is correct ?

**Comment by nappex** [comment]
All possible types of content types:  https://www.iana.org/assignments/media-types/media-types.xhtml

**Comment by encukou** [comment]
> A generic version of collections.abc.Iterator

"Generic" is a type-hinting term, see e.g. https://docs.python.org/3/library/stdtypes.html#types-genericalias – basically, generic types can have another type associated with them, like a collection with typed contents (`List[int]`) or an iterator that yields a specific type (`Iterator[str]`).

> why do we use var `file_path`, when we don't use it in code ?

It's there because tuple unpacking is clearer than getting only the second item (with `[1]`).
It's named according to what's in it. If we ever need it in the future, we can simply start using it.

`_` is good when you don't want to think of a name, but I don't think all unused vars should be named `_`. Should they?

**Comment by nappex** [comment]
How the environ does create headers ?

**Comment by nappex** [comment]
How to test:
1. create app with content html and css - result links will be only html or css
2. create app with any content - result as no links
3. Better way should be - no app, but parametrize test for scanning func and paramaterize args will be (all args, one of each, no links)
there are problems:
- how to get file as input to scanning func (app, parametrize test)
- how to get result to compare (parametrize test)

**Comment by encukou** [comment]
> How the environ does create headers?

In Flask, you usually return a string from a view function:

```
@app.route('/'):
def index():
    return "content of webpage"
```
Flask then takes the string and builds a  [`Response`](https://flask.palletsprojects.com/en/2.0.x/api/#flask.Response) object.
(You can also return something else than a string; see the types [`make_response`](https://flask.palletsprojects.com/en/2.0.x/api/#flask.Flask.make_response) accepts.)

But you can also return a [`Response`](https://flask.palletsprojects.com/en/2.0.x/api/#flask.Response) directly, and set the headers there.
Note that the the `Content-Type` header can contain a mimetype and encoding, like `text/html; charset=utf-8`. Use `mimetype="text/html"` and let Flask build the full header for you.

> how to get file as input to scanning func (app, parametrize test)

Use [`io.BytesIO(b'contents')`](https://docs.python.org/3/library/io.html#io.BytesIO) to get an object that acts like an open file. (`freezeyt` wants bytes, but if you ever want a file-like object opened in text mode, there's also [`StringIO`](https://docs.python.org/3/library/io.html#io.StringIO).)

> how to get result to compare (parametrize test)

I'm not sure I understand the question. You should know the expected result for each set of parameters.

**Comment by nappex** [comment]
What is right way to declare type of variable ?

- use built object as `str`, `dict`, `list` or
- from typing import `Dict`, `List`, `Str`

**Comment by encukou** [comment]
Built-in objects are nicer, since you don't need to import anything. However, generics like `list[str]` are only available in Python 3.9+, so for that you need `typing.List[str]`.
Also consider if you strictly need `list`/`dict` or if [Iterable](https://docs.python.org/3/library/typing.html#typing.Iterable), [Sequence](https://docs.python.org/3/library/typing.html#typing.Sequence) or [Mapping](https://docs.python.org/3/library/typing.html#typing.Mapping) will do. (This is mentioned in the [typing docs](https://docs.python.org/3/library/typing.html#corresponding-to-built-in-types)).

**Comment by nappex** [comment]
It seems that my approach hit to tests of the cli. I do not think that we wanna insert configuration of scanners to terminal, so two ways of solve I have in my mind, 1) create app just for testing cli or 2) set up default values (just for html and css) if scanners configuration missing and when we would like to turn off we have to specify which of defaults should turn off.

What do you think ?

**Comment by encukou** [comment]
There should definitely be a good default (`html` and `css`).

I would say that for now, if the user specifies `url_finders` the default should be completely overridden. So you can set `url_finders: {}` to turn everything off, but to add another one you'd need to repeat the default, e.g.:

    url_finders:
        text/html: get_html_links
        text/css: get_css_links
        image/svg+xml: mymodule:get_svg_links

That's not the ideal behaviour, but it's not too bad and should be easy to implement.

File: tests/test_config.py [hunk 1]
```diff
@@ -133,3 +136,65 @@ def gen(app):
...
+    assert (builddir / 'static' / 'OFL.txt').exists()
+    assert not (builddir / 'static' / 'style.css').exists()
+    assert not (builddir / 'static' / 'TurretRoad-Regular.ttf').exists()
+
+
+def test_get_no_links(tmp_path):
+    """Test if we get only links from css files.
+    We used app with css (app_links_css) and extra pages.
+    Extra page is preserved.
+    """
```
Comment (encukou) [inline tests/test_config.py:189]: <kbd>ctrl</kbd>+<kbd>v</kbd>? :)

File: freezeyt/url_finders.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+from werkzeug.datastructures import Headers
+from werkzeug.http import parse_options_header
+
+from freezeyt.encoding import decode_input_path
+
+
+def get_css_links(
+    css_file: bytes, base_url: str, headers: Headers = None
+)  -> List[str]:
```
Comment (encukou) [inline freezeyt/url_finders.py:16]: * Dal bych `-> Iterable[str]`; tuhle funkci kdykoli můžeme třeba změnit na generátor.

File: freezeyt/freezer.py [hunk 3]
```diff
@@ -369,7 +401,8 @@ def handle_redirects(self):
         for task in self.redirecting_tasks.values():
             if task.redirects_to.status != TaskStatus.DONE:
                 raise InfiniteRedirection(
-                    f'{task.get_a_url()} redirects to {task.redirects_to.get_a_url()}, which was not frozen (most likely because of infinite redirection)'
+                    f'{task.get_a_url()} redirects to {task.redirects_to.get_a_url()},'
+                    + 'which was not frozen (most likely because of infinite redirection)'
```
Comment (encukou) [inline freezeyt/freezer.py:405]: Pokročilá verze úkolu z první lekce :)
image

Comment (nappex) [reply freezeyt/freezer.py:405]: Není to chyba, jen jsem chtěl zkrátit řádek přišlo mi to dlouhé...

File: tests/test_config.py (hunk 1 above)
Comment (nappex) [reply tests/test_config.py:189]: jj ctrl+v, s tím, že si testy nejsem moc jistý a myslím si že se budou muset předělat

File: freezeyt/url_finders.py (hunk 2 above)
Comment (nappex) [reply freezeyt/url_finders.py:16]: jasně díky na to jsem zapomněl "obecnější"

File: freezeyt/freezer.py (hunk 3 above)
Comment (encukou) [reply freezeyt/freezer.py:405]: Ale chybí tam mezera po čárce.

Comment (nappex) [reply freezeyt/freezer.py:405]: jo takhle... Já to pochopil, že to myslíš proč jsem to opravoval co tam bylo za chybu..

File: tests/test_import_variable_from_module.py [hunk 4]
```diff
@@ -38,11 +38,55 @@ def test_missing_module():
...
+def test_overridden_default_module():
+    joinpath = import_variable_from_module(
+        "pathlib:Path.joinpath", default_module='os'
+    )
+    assert joinpath.__name__ == 'joinpath'
+
+
+def test_default_module():
+    joinpath = import_variable_from_module("Path.joinpath", default_module='pathlib')
+    assert joinpath.__name__ == 'joinpath'
```
Comment (nappex) [inline tests/test_import_variable_from_module.py:87]: to review

**Review by nappex — CHANGES_REQUESTED** [review]

File: freezeyt/freezer.py [hunk 5]
```diff
@@ -36,17 +34,41 @@ def freeze(app, config):
...
+    for content_type, url_finder in url_finders.items():
+        if isinstance(url_finder, str):
+            url_finder = import_variable_from_module(
+                url_finder, default_module='freezeyt.url_finders'
+            )
+        elif not hasattr(url_finder, '__call__'):
+            raise TypeError(
+                "Url-finder in configuration must be a string or a callable,"
+                + f" not {type(url_finder)}!"
+            )
```
Comment (encukou) [inline freezeyt/freezer.py:67]:
```suggestion
        elif not callable(url_finder):
            raise TypeError(
                "Url-finder for {content_type!r} in configuration must be a string or a callable,"
                + f" not {type(url_finder)}!"
            )
```

File: freezeyt/url_finders.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import css_parser
+
+from werkzeug.datastructures import Headers
+from werkzeug.http import parse_options_header
+
+from freezeyt.encoding import decode_input_path
+
+
+def get_css_links(
+    css_file: BinaryIO, base_url: str, headers: List[Tuple[str]]=None
```
Comment (encukou) [inline freezeyt/url_finders.py:15]:
```suggestion
    css_file: BinaryIO, base_url: str, headers: List[Tuple[str, str]]=None
```

File: freezeyt/url_finders.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    text = css_file.read()
+    parsed = css_parser.parseString(text)
+    all_urls = css_parser.getUrls(parsed)
+    for url in all_urls:
+        result.append(urljoin(base_url, url))
+    return result
+
+
+def get_html_links(
+    page_content: bytes, base_url: str, headers: List[Tuple[str]]=None
```
Comment (encukou) [inline freezeyt/url_finders.py:28]:
```suggestion
    page_content: bytes, base_url: str, headers: List[Tuple[str, str]]=None
```

File: tests/test_import_variable_from_module.py (hunk 4 above)
Comment (encukou) [reply tests/test_import_variable_from_module.py:87]: looks great!

File: tests/test_url_finders.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+from freezeyt.url_finders import get_html_links
+from freezeyt.url_finders import get_css_links
+from testutil import context_for_test
+
+
+TEST_DATA = {
+    'html_links_fallback':
+        (   'text/html',
+            {'text/html': 'get_html_links'},
+            ('freezeyt.url_finders', 'get_html_links', get_html_links)
```
Comment (encukou) [inline tests/test_url_finders.py:14]:
```suggestion
            get_html_links,
```
No need to test `__module__` and `__name__` here.

File: tests/test_url_finders.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            **module.freeze_config,
+            'output': str(builddir),
+            'url_finders': {'text/css': 'get_css_links'},
+        }
+
+        freeze(module.app, freeze_config)
+
+    assert (builddir / 'index.html').exists()
+    assert (builddir / 'static' / 'OFL.txt').exists()
+    assert not (builddir / 'static' / 'style.css').exists()
```
Comment (encukou) [inline tests/test_url_finders.py:79]: Put this in extra_pages.

File: tests/test_url_finders.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+    assert (builddir / 'index.html').exists()
+    assert not (builddir / 'second_page.html').exists()
+
+
+def test_get_url_finder_callable_defined_by_user(tmp_path):
+    """Test if we freezer parse url_finder inserted as func type by user.
+    """
+    def my_url_finder(page_content, base_url, headers=None):
+        return []
```
Comment (encukou) [inline tests/test_url_finders.py:105]: Try with a finder that finds something.

File: freezeyt/url_finders.py [hunk 11]
```diff
@@ -12,20 +12,23 @@
...
-    return result
+        yield urljoin(base_url, url)
+
 
 
 def get_html_links(
-    page_content: bytes, base_url: str, headers: List[Tuple[str]]=None
+    page_content: bytes,
+    base_url: str,
+    headers: Optional[List[Tuple[str, str]]]=None
```
Comment (encukou) [inline freezeyt/url_finders.py:31]: Tady se na konec řádku dává čárka navíc, aby se tenhle řádek nezměnil když se přidá další argument.

File: freezeyt/url_finders.py [hunk 12]
```diff
@@ -34,7 +37,9 @@ def get_html_links(
     base_url is the URL of the page.
     """
 
-    def get_links_from_node(node: xml.dom.minidom.Node, base_url) -> list:
+    def get_links_from_node(
+        node: xml.etree.ElementTree.Element, base_url: str
```
Comment (encukou) [inline freezeyt/url_finders.py:41]: Je to prkotina, ale zajímalo by mě proč i tohle nerozhodíš na dva řádky? Na první pohled to vypadá jako jeden argument s komplikovaným typem.

Comment (nappex) [reply freezeyt/url_finders.py:41]: klíč podle kterého zalamuji já je, že pokud se řádek nevleze do 79 znaků, tak se ho snažím smysluplně zalomit. většinou tak, že celý řádek odentruju o řádek níže, pokud se i potom nevleze na 79 znaků, tak jej většinou rozhodím, na každý řádek jeden parametr nebo také pokud je nějaký jeden hodně dlouhy parametr, tak jak je to  v tomto případě u všech ostatních. V tomto případě je to také dlouhý parametr - a proč jsem jej nechal takto? Popravdě nevím, možná, že jsem spěchal a odflákl nebo občas se mi stává, že jak programuji v každé volné chvíli a mám to programování hodně rozkouskováno tak na některé věci zapomenu nebo se pořádně nezamyslím.

Určitě mi nevadí, že mi opravuješ prkotiny rád bych psal kód tak, jak by se psát měl.

File: freezeyt/url_finders.py (hunk 11 above)
Comment (nappex) [reply freezeyt/url_finders.py:31]: ok nebyl jsem si jistý

File: tests/test_url_finders.py [hunk 13]
```diff
@@ -72,12 +70,13 @@ def test_get_only_links_from_css(tmp_path):
...
 
+        freeze_config['extra_pages'] += ["/static/style.css"]
         freeze(module.app, freeze_config)
 
     assert (builddir / 'index.html').exists()
     assert (builddir / 'static' / 'OFL.txt').exists()
-    assert not (builddir / 'static' / 'style.css').exists()
-    assert not (builddir / 'static' / 'TurretRoad-Regular.ttf').exists()
+    assert (builddir / 'static' / 'style.css').exists()
+    assert (builddir / 'static' / 'TurretRoad-Regular.ttf').exists()
```
Comment (encukou) [inline tests/test_url_finders.py:79]: Now there's no page that didn't match, so the result is the same as if all links were followed...

Comment (nappex) [reply tests/test_url_finders.py:79]: It means to update whole `app_links_css`

Comment (encukou) [reply tests/test_url_finders.py:79]: Yes. (Or create a new app, but that would probably be more work.)

**Review by nappex — APPROVED** [review]

**Pull request merged**